Tests/SwiftGRPCTests/ChannelShutdownTests.swift

```diff
@@ -0,0 +1,97 @@
+ */
+@testable import SwiftGRPC
+import XCTest
+
+final class ChannelShutdownTests: BasicEchoTestCase {
+  static var allTests: [(String, (ChannelShutdownTests) -> () throws -> Void)] {
+    return [
+      ("testThrowsWhenCreatingCallWithAlreadyShutDownChannel", testThrowsWhenCreatingCallWithAlreadyShutDownChannel),
+      ("testCallReceiveThrowsWhenChannelIsShutDown", testCallReceiveThrowsWhenChannelIsShutDown),
+      ("testCallCloseThrowsWhenChannelIsShutDown", testCallCloseThrowsWhenChannelIsShutDown),
+      ("testCallCloseAndReceiveThrowsWhenChannelIsShutDown", testCallCloseAndReceiveThrowsWhenChannelIsShutDown),
+      ("testCallSendThrowsWhenChannelIsShutDown", testCallSendThrowsWhenChannelIsShutDown),
+      ("testCancelsActiveCallWhenShutdownIsCalled", testCancelsActiveCallWhenShutdownIsCalled),
+    ]
+  }
+}
+
+extension ChannelShutdownTests {
+  func testThrowsWhenCreatingCallWithAlreadyShutDownChannel() {
+    self.client.channel.shutdown()
+
+    XCTAssertThrowsError(try self.client.channel.makeCall("foobar")) { error in
+      XCTAssertEqual(.alreadyShutdown, error as? Channel.Error)
+    }
+  }
+
+  func testCallReceiveThrowsWhenChannelIsShutDown() {
+    let call = try! self.client.channel.makeCall("foo")
+    self.client.channel.shutdown()
+
+    XCTAssertThrowsError(try call.receiveMessage(completion: { _ in })) { error in
+      XCTAssertEqual(.completionQueueShutdown, error as? CallError)
+      XCTAssertNotNil(self.client.channel)
+    }
+  }
+
+  func testCallCloseThrowsWhenChannelIsShutDown() {
+    let call = try! self.client.channel.makeCall("foo")
+    self.client.channel.shutdown()
+
+    XCTAssertThrowsError(try call.close()) { error in
+      XCTAssertEqual(.completionQueueShutdown, error as? CallError)
+      XCTAssertNotNil(self.client.channel)
+    }
+  }
+
+  func testCallCloseAndReceiveThrowsWhenChannelIsShutDown() {
+    let call = try! self.client.channel.makeCall("foo")
+    self.client.channel.shutdown()
+
+    XCTAssertThrowsError(try call.closeAndReceiveMessage(completion: { _ in })) { error in
+      XCTAssertEqual(.completionQueueShutdown, error as? CallError)
+      XCTAssertNotNil(self.client.channel)
+    }
+  }
+
+  func testCallSendThrowsWhenChannelIsShutDown() {
+    let call = try! self.client.channel.makeCall("foo")
+    self.client.channel.shutdown()
+
+    XCTAssertThrowsError(try call.sendMessage(data: Data())) { error in
+      XCTAssertEqual(.completionQueueShutdown, error as? CallError)
+      XCTAssertNotNil(self.client.channel)
+    }
+  }
+
+  func testCancelsActiveCallWhenShutdownIsCalled() {
+    let errorExpectation = self.expectation(description: "error is returned to call when channel is shut down")
+    let call = try! self.client.channel.makeCall("foo")
+
+    try! call.receiveMessage { result in
+      XCTAssertFalse(result.success)
+      errorExpectation.fulfill()
+    }
+
+    self.client.channel.shutdown()
+    self.waitForExpectations(timeout: 0.1)
```

Comment:
This is less than the regular call timeout, ensuring that the channel shutdown shuts down the call _before it would have been finished due to a timeout, anyway_, right? Would this still pass if you remove `self.activeCalls.forEach { $0.value?.cancel() }` from `channel.shutdown()`? In that case, maybe we could get rid of `Channel.activeCalls` altogether?

Reply:
>  This is less than the regular call timeout, ensuring that the channel shutdown shuts down the call _before it would have been finished due to a timeout, anyway_, right?

Yes, correct

>  Would this still pass if you remove `self.activeCalls.forEach { $0.value?.cancel() }` from `channel.shutdown()`?

Nope, commenting out that line causes this test to fail

Reply:
Thanks for checking!